File: packages/terminal/android/src/main/java/com/getcapacitor/community/stripe/terminal/StripeTerminalPlugin.java
package com.getcapacitor.community.stripe.terminal;

import android.Manifest;
import android.os.Build;
import android.util.Log;
import androidx.annotation.RequiresApi;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;
import com.stripe.stripeterminal.external.models.TerminalException;
import java.util.Objects;

@RequiresApi(api = Build.VERSION_CODES.S)
@CapacitorPlugin(
    name = "StripeTerminal",
    permissions = {
        @Permission(alias = "location", strings = { Manifest.permission.ACCESS_FINE_LOCATION }),
        @Permission(alias = "bluetooth_old", strings = { Manifest.permission.BLUETOOTH, Manifest.permission.BLUETOOTH_ADMIN }),
        @Permission(
            alias = "bluetooth",
            strings = { Manifest.permission.BLUETOOTH_SCAN, Manifest.permission.BLUETOOTH_CONNECT, Manifest.permission.BLUETOOTH_ADVERTISE }
        )
    }
)
public class StripeTerminalPlugin extends Plugin {

    private final StripeTerminal implementation = new StripeTerminal(
        this::getContext,
        this::getActivity,
        this::notifyListeners,
        getLogTag()
    );

    @PluginMethod
    public void initialize(PluginCall call) throws TerminalException {
        this._initialize(call);
    }

    @PluginMethod
    public void setConnectionToken(PluginCall call) {
        this.implementation.setConnectionToken(call);
    }

    @PluginMethod
    public void setSimulatorConfiguration(PluginCall call) {
        this.implementation.setSimulatorConfiguration(call);
    }

    @PermissionCallback
    private void locationPermsCallback(PluginCall call) throws TerminalException {
        if (getPermissionState("location") == PermissionState.GRANTED) {
            this._initialize(call);
        } else {
            requestPermissionForAlias("location", call, "locationPermsCallback");
        }
    }

    @PermissionCallback
    private void bluetoothOldPermsCallback(PluginCall call) throws TerminalException {
        if (getPermissionState("bluetooth_old") == PermissionState.GRANTED) {
            if (Objects.equals(call.getMethodName(), "discoverReaders")) {
                this.discoverReaders(call);
            } else {
                this.connectReader(call);
            }
        } else {
            requestPermissionForAlias("bluetooth_old", call, "bluetoothOldPermsCallback");
        }
    }

    @PermissionCallback
    private void bluetoothPermsCallback(PluginCall call) throws TerminalException {
        if (getPermissionState("bluetooth") == PermissionState.GRANTED) {
            if (Objects.equals(call.getMethodName(), "discoverReaders")) {
                this.discoverReaders(call);
            } else {
                this.connectReader(call);
            }
        } else {
            requestPermissionForAlias("bluetooth", call, "bluetoothPermsCallback");
        }
    }

    private void _initialize(PluginCall call) throws TerminalException {
        if (getPermissionState("location") != PermissionState.GRANTED) {
            requestPermissionForAlias("location", call, "locationPermsCallback");
        } else {
            Log.d("Capacitor:permission location", getPermissionState("location").toString());
            this.implementation.initialize(call);
        }
    }

    @PluginMethod(returnType =  PluginMethod.RETURN_CALLBACK)
    public void discoverReaders(PluginCall call) {
        if (
            Objects.equals(call.getString("type"), TerminalConnectTypes.Bluetooth.getWebEventName()) ||
            Objects.equals(call.getString("type"), TerminalConnectTypes.Simulated.getWebEventName())
        ) {
            Log.d("Capacitor:permission bluetooth_old", getPermissionState("bluetooth_old").toString());
            Log.d("Capacitor:permission bluetooth", getPermissionState("bluetooth").toString());
            if (Build.VERSION.SDK_INT <= 30 && getPermissionState("bluetooth_old") != PermissionState.GRANTED) {
                requestPermissionForAlias("bluetooth_old", call, "bluetoothOldPermsCallback");
            } else if (Build.VERSION.SDK_INT > 30 && getPermissionState("bluetooth") != PermissionState.GRANTED) {
                requestPermissionForAlias("bluetooth", call, "bluetoothPermsCallback");
            } else {
                this.implementation.onDiscoverReaders(call);
            }
        } else {
            this.implementation.onDiscoverReaders(call);
        }
    }

    @PluginMethod
    public void cancelDiscoverReaders(PluginCall call) {
        this.implementation.cancelDiscoverReaders(call);
    }

    @PluginMethod
    public void connectReader(PluginCall call) {
        if (Objects.equals(call.getString("type"), TerminalConnectTypes.Bluetooth.getWebEventName())) {
            Log.d("Capacitor:permission bluetooth_old", getPermissionState("bluetooth_old").toString());
            Log.d("Capacitor:permission bluetooth", getPermissionState("bluetooth").toString());
            if (Build.VERSION.SDK_INT <= 30 && getPermissionState("bluetooth_old") != PermissionState.GRANTED) {
                requestPermissionForAlias("bluetooth_old", call, "bluetoothOldPermsCallback");
            } else if (Build.VERSION.SDK_INT > 30 && getPermissionState("bluetooth") != PermissionState.GRANTED) {
                requestPermissionForAlias("bluetooth", call, "bluetoothPermsCallback");
            } else {
                this.implementation.connectReader(call);
            }
        } else {
            this.implementation.connectReader(call);
        }
    }

    @PluginMethod
    public void getConnectedReader(final PluginCall call) {
        this.implementation.getConnectedReader(call);
    }

    @PluginMethod
    public void disconnectReader(final PluginCall call) {
        this.implementation.disconnectReader(call);
    }

    @PluginMethod
    public void collectPaymentMethod(PluginCall call) {
        this.implementation.collectPaymentMethod(call);
    }

    @PluginMethod
    public void cancelCollectPaymentMethod(final PluginCall call) {
        this.implementation.cancelCollectPaymentMethod(call);
    }

    @PluginMethod
    public void confirmPaymentIntent(PluginCall call) {
        this.implementation.confirmPaymentIntent(call);
    }

    @PluginMethod
    public void installAvailableUpdate(PluginCall call) {
        this.implementation.installAvailableUpdate(call);
    }

    @PluginMethod
    public void cancelInstallUpdate(PluginCall call) {
        this.implementation.cancelInstallUpdate(call);
    }

    @PluginMethod
    public void setReaderDisplay(PluginCall call) {
        this.implementation.setReaderDisplay(call);
    }

    @PluginMethod
    public void clearReaderDisplay(PluginCall call) {
        this.implementation.clearReaderDisplay(call);
    }

    @PluginMethod
    public void rebootReader(PluginCall call) {
        this.implementation.rebootReader(call);
    }

    @PluginMethod
    public void cancelReaderReconnection(PluginCall call) {
        this.implementation.cancelReaderReconnection(call);
    }
}

File: packages/terminal/ios/Sources/StripeTerminalPlugin/StripeTerminalPlugin.swift
import Foundation
import StripeTerminal
import Capacitor
import PassKit

@objc(StripeTerminalPlugin)
public class StripeTerminalPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "StripeTerminalPlugin"
    public let jsName = "StripeTerminal"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setConnectionToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "discoverReaders", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelDiscoverReaders", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "connectReader", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getConnectedReader", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnectReader", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "collectPaymentMethod", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelCollectPaymentMethod", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "confirmPaymentIntent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setSimulatorConfiguration", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "installAvailableUpdate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelInstallUpdate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setReaderDisplay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearReaderDisplay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "rebootReader", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelReaderReconnection", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = StripeTerminal()

    override public func load() {
        super.load()
        self.implementation.plugin = self
        // TODO: add STPAPIClient.shared.appInfo
    }

    @objc func initialize(_ call: CAPPluginCall) {
        self.implementation.initialize(call)
    }

    @objc func setConnectionToken(_ call: CAPPluginCall) {
        self.implementation.setConnectionToken(call)
    }

    @objc func discoverReaders(_ call: CAPPluginCall) {
        do {
            try self.implementation.discoverReaders(call)
        } catch {
            call.reject("discoverReaders throw error.")
        }
    }

    @objc func setSimulatorConfiguration(_ call: CAPPluginCall) {
        self.implementation.setSimulatorConfiguration(call)
    }

    @objc func cancelDiscoverReaders(_ call: CAPPluginCall) {
        self.implementation.cancelDiscoverReaders(call)
    }

    @objc func connectReader(_ call: CAPPluginCall) {
        self.implementation.connectReader(call)
    }

    @objc func getConnectedReader(_ call: CAPPluginCall) {
        self.implementation.getConnectedReader(call)
    }

    @objc func disconnectReader(_ call: CAPPluginCall) {
        self.implementation.disconnectReader(call)
    }

    @objc func collectPaymentMethod(_ call: CAPPluginCall) {
        self.implementation.collectPaymentMethod(call)
    }

    @objc func cancelCollectPaymentMethod(_ call: CAPPluginCall) {
        self.implementation.cancelCollectPaymentMethod(call)
    }

    @objc func confirmPaymentIntent(_ call: CAPPluginCall) {
        self.implementation.confirmPaymentIntent(call)
    }

    @objc func installAvailableUpdate(_ call: CAPPluginCall) {
        self.implementation.installAvailableUpdate(call)
    }

    @objc func cancelInstallUpdate(_ call: CAPPluginCall) {
        self.implementation.cancelInstallUpdate(call)
    }

    @objc func setReaderDisplay(_ call: CAPPluginCall) {
        self.implementation.setReaderDisplay(call)
    }

    @objc func clearReaderDisplay(_ call: CAPPluginCall) {
        self.implementation.clearReaderDisplay(call)
    }

    @objc func rebootReader(_ call: CAPPluginCall) {
        self.implementation.rebootReader(call)
    }

    @objc func cancelReaderReconnection(_ call: CAPPluginCall) {
        self.implementation.cancelReaderReconnection(call)
    }

}
